okapi/core/OkapiServiceRunner.php

```diff
@@ -42,6 +42,7 @@ class OkapiServiceRunner
         'services/caches/formatters/gpx',
         'services/caches/formatters/garmin',
         'services/caches/formatters/ggz',
+        'services/fieldnotes/upload',
```

Comment:
I wonder if it wouldn't be easier to add a new parameter to the `submit` method instead - which, if given, would insert the log entry to the "draft" table, instead of the final table. This would allow the devs to use an existing method which they're already familiar to. On the other hand, it would not support the "fieldnotes" format. Are you confident devs prefer to this format over JSON?

Reply:
The first part, with respect to selecting a better namespace is ok with me.

The second, merging it into existint log methods is not, at least not yet. I do not yet have sufficient insight into OKAPI in order to have a position here. I can only assume, when you are making such a suggestion that merging it into the submit method would save us from duplicating functionality. Functionality that is already available in that method and we would simply reuse it? You are further assuming that adding if then else conditions into the existing function and targetting different database tables is the lesser evil? One major motivation for me to not touch existing methods is to guarantee that I will not break existing functions. Also I am quite sure there are client applications out there consuming OKAPI, that are no longer actively maintained and it may be a show stopper for those if all of a suden a method behaves only slightly different than it did in the past. It is a problem even it is actually a bug in OKAPI as there is nobody out there chasing it down to OKAPI, some appications will just silently no longer work.

I'd say: look at the PR's code and find out whether the logic is complex and could be simplified by moving stuff into the submit method. I'd clearly need help in doing so as those functions aren't really simple. But that should not keep us from not doing it, if it makes sense for other reasons.

I can only describe what I had in mind with those fieldrecords in general and let us clean up the name space a bit here:

`draftlogs/upload - we have discussed this`
`draftlogs/query  - return a list of ids to later target the methods below`
`draftlogs/flush - remove all draft logs from the database`
`draftlogs/delete - delete a specific draft log based on what query returned`

Also keep in mind that the `upload` operates on potentially very large arrays. It is not uncommon that such fieldnotes files have hunderts of entries. Throwing a particular record into the right database isn't the challenge here. The challenge lies in parsing the input data appropriaetely. Well, while I am writing this, is your thinking perhaps to leave ll the parsing as it is and at the point we deal with the database we'd call an internal service such as this one:
```
$geocache = OkapiServiceRunner::call(
            'services/caches/geocache',
            new OkapiInternalRequest($request->consumer, $request->token, array(
                'cache_code' => $cache_code,
                'fields' => 'my_notes|internal_id'
            ))
        );

```
... of course for `logs/submit` then?

Reply:
I suggested that because I suspect it would be easier for external developers, API-wise. Imagine an app which already is able to submit log entries, and imagine the developer wants to add a "save as draft instead" checkbox to this app. In such case, adding an argument to the existing `submit` method would be a simple way of providing them with such functionality.

It does have many drawbacks however - e.g. it would obviously be more difficult to implement on our side, and we some arguments of the `submit` method would not work (especially given that OCDE keeps draftlogs and regular logs in separate tables). I also agree that you probably should not touch any existing methods until you're able to test them on OCPL database.

So to sum up, I agree it's good to keep it separate. +1 for creating the `draftlogs` directly, but let's stick with existing naming conventions for the methods (e.g. we don't have `query` methods, but we have `geocache` and `geocaches`), so this would be:

```
services/draftlogs/upload_fieldnotes_file
services/draftlogs/draftlog (retrieves one)
services/draftlogs/draftlogs (retrieves many)
services/draftlogs/add (JSON alternative to use instead upload_fieldnotes_file for better error control etc.)
services/draftlogs/delete
```

Reply:
All good what you're suggesting. I am also with you with respect to the JSON except that no single client application exists that would use it and just like so many other applications most of these are virtually dead. cgeo is the only one that really lives. We need to discuss this with @Lineflyer whether thinks cgeo would want to use the JSON path,

I will update the PR accordingly. e.g. moving stuff into the right namespace and I will leave it with just the upload for now as cgeo needs the upload. They would never use any of the other methods.

Reply:
@hxdimpf I am probably the wrong person to discuss "coding stuff" as I am no programmer myself but only a bit responsible for support and issue sorting and alike in c:geo.
Best is to ask your question or your request to c:geo directly in a matching issue on the c:geo Github page.